feat: enable soft delete for resources

haritianaadriano opened the pull request.

haritianaadriano committed:
feat: enable soft delete for resources

Mahefaa commented on src/main/resources/db/migration/V42_10__Update_resources_data.sql:
resources is too vague as a name

Mahefaa commented on src/main/java/school/hei/haapi/repository/PaymentRepository.java:
```diff
@@ -13,7 +13,8 @@ public interface PaymentRepository extends JpaRepository<Payment, String> {
           "select p from Payment p join Fee f on f.id = p.fee.id"
-              + " where f.student.id = :student_id and f.id = :fee_id")
+              + " where f.student.id = :student_id and f.id = :fee_id"
+              + " and p.isDeleted = false")
```
test without adding isDeleted check in the sql request.

Normally the @Where decorator already checks this field's value

Mahefaa commented on src/main/java/school/hei/haapi/repository/dao/CourseDao.java:
```diff
@@ -68,6 +68,8 @@ public List<Course> findByCriteria(
               builder.lower(teacher.get("firstName")), "%" + teacherFirstName.toLowerCase() + "%"));
     }
 
+    predicates.add(builder.isFalse(root.get("isDeleted")));
```
not needed if the @Where decorator does it but test it anyway

Mahefaa commented on src/main/java/school/hei/haapi/repository/dao/UserManagerDao.java:
```diff
@@ -50,6 +50,8 @@ public List<User> findByCriteria(
 
     Predicate hasUserRole = builder.equal(root.get("role"), role);
 
+    Predicate isNotDeleted = builder.isFalse(root.get("isDeleted"));
+
```
same for this

Mahefaa commented on src/main/resources/db/migration/V42_10__Update_resources_data.sql:
```diff
@@ -0,0 +1,14 @@
+-- alter table user
+alter table "user" add column if not exists is_deleted boolean default false;
+
+-- alter table payment
+alter table "payment" add column if not exists is_deleted boolean default false;
```
delete payment ? ... seems a bit dubious but ok

Mahefaa commented on src/main/java/school/hei/haapi/model/Course.java:
```diff
@@ -34,6 +38,8 @@ public class Course implements Serializable {
 
   private Integer totalHours;
 
+  private Boolean isDeleted;
```
why "Boolean" instead of boolean ? your fields  default to false !

Mahefaa reviewed (changes requested):
NO !

for a soft delete like this, adding a boolean is indeed the right way but add

haritianaadriano committed:
refactor: use default where when getting data

haritianaadriano committed:
feat: enable soft delete

haritianaadriano committed:
test: manager can delete some resource

haritianaadriano committed:
fix: manager can delete teacher

Mahefaa commented on doc/api.yml:
```diff
@@ -233,6 +233,34 @@ paths:
+            type: string
+      responses:
+        '200':
+          description: student deleted.
```
deleted student

Mahefaa commented on doc/api.yml:
```diff
@@ -233,6 +233,34 @@ paths:
+    delete:
+      tags:
+        - Users
+      summary: Delete student by identifier.
```
describe it as a soft delete

Mahefaa commented on doc/api.yml:
```diff
@@ -1649,6 +1767,45 @@ paths:
+    delete:
+      tags:
+        - Paying
+      summary: Delete student payment by identifier.
```
delete student payment on a fee

Mahefaa commented on doc/api.yml:
```diff
@@ -1649,6 +1767,45 @@ paths:
+      tags:
+        - Paying
+      summary: Delete student payment by identifier.
+      operationId: deleteStudentPaymentById
```
deleteStudentFeePaymentById

Mahefaa commented on src/main/java/school/hei/haapi/model/Course.java:
```diff
@@ -11,6 +11,8 @@
 import javax.persistence.OneToMany;
 import javax.persistence.Table;
 import lombok.*;
```
stop using import *

Mahefaa commented on src/main/java/school/hei/haapi/model/Fee.java:
```diff
@@ -72,6 +76,8 @@ public class Fee implements Serializable {
   @OneToMany(mappedBy = "fee")
   private List<Payment> payments;
 
+  private boolean isDeleted = false;
```
why ?

Mahefaa commented on src/main/java/school/hei/haapi/service/CourseService.java:
```diff
@@ -50,6 +50,12 @@ public List<Course> getCourses(
   }
 
+  public Course deleteById(String courseId) {
+    Course course = getById(courseId);
```
why do you need to get it before deleting it ??

haritianaadriano committed:
test: assert that the deleted resources is not return

test: refactor by using assertFalse

Mahefaa commented on src/test/java/school/hei/haapi/integration/FeeIT.java:
```diff
@@ -59,6 +59,19 @@ void setUp() {
+    assertEquals(fee1(), deletedFee);
+
+    List<Fee> fees = api.getStudentFees(STUDENT1_ID, 1, 5, null);
+    assertTrue(!fees.contains(deletedFee));
```
use assertFalse

Mahefaa commented on src/main/java/school/hei/haapi/service/GroupService.java:
```diff
@@ -28,6 +28,12 @@ public class GroupService {
+    return group;
+  }
+
   public Group getById(String groupId) {
     return repository.getById(groupId);
   }
```
use findBy and orElseThrow

haritianaadriano replied:
but it's not redundant ?

Mahefaa commented on src/test/java/school/hei/haapi/integration/StudentIT.java:
```diff
@@ -278,6 +279,27 @@ void student_update_own_ok() throws ApiException {
   }
 
+  @Test
+  void student_delete_other_ko() throws ApiException {
+    ApiClient student1Client = anApiClient(STUDENT1_TOKEN);
+    UsersApi api = new UsersApi(student1Client);
+
+    assertThrowsForbiddenException(() -> api.deleteStudentById(STUDENT3_ID));
+  }
```
student _delete_ko whether its their id or not

Mahefaa reviewed (changes requested):
assertFalse exists x)

haritianaadriano committed:
chore: handle if given id doesn't exist

chore: handle parent delete for children resources

docs: soft delete as description

Mahefaa commented on doc/api.yml:
```diff
@@ -1102,6 +1160,34 @@ paths:
+    delete:
+      tags:
+        - Users
+      summary: Delete teacher by identifier.
+      operationId: deleteTeacherById
```
description: soft delete

Mahefaa commented on src/main/java/school/hei/haapi/model/Course.java:
```diff
@@ -34,6 +44,8 @@ public class Course implements Serializable {
 
   private Integer totalHours;
 
+  private boolean isDeleted = false;
```
why = false !

haritianaadriano replied:
you said that we need to set it to false by default

Mahefaa reviewed (changes requested):
what will happen to a student's resources if the student gets deleted ?

haritianaadriano committed:
test: handle removing parent table

test: handle remove user linked with fee

temp

Mahefaa commented on doc/api.yml:
```diff
@@ -404,6 +433,34 @@ paths:
+    delete:
+      tags:
+        - Teaching
+      summary: Delete course by identifier.
+      operationId: deleteCourseById
```
description for soft delete

Mahefaa commented on src/main/java/school/hei/haapi/model/GroupFlow.java:
```diff
@@ -36,18 +42,22 @@
   private String id;
 
-  @ManyToOne(fetch = LAZY)
+  @ManyToOne(fetch = LAZY, cascade = REMOVE)
   @JoinColumn(name = "student_id")
+  @OnDelete(action = CASCADE)
```
Deleting a groupFlow will delete the related student ??

Mahefaa reviewed (changes requested):
it definitely doesn't tell me what happens to child entities on parent entity removal.

Also, have you tested whether the is_deleted variable really changes ?

haritianaadriano committed:
chore: handle removing parent entity
